Updated linting to ruff

RabiyaF opened the pull request:
<!---

Provide a short summary of this PR in the title above.
This will be used to generate release note, so please write this in the
past tense and use language that should be understandable to a potential user.

-->

## Description of work

Fixes #1303

- [x] The linting now uses `Ruff`.
- [x] `pylint`, `flake8` and `isort`  have been removed from the dev dependencies or precommit hooks.
- [x] docs have been updated.
- [x] Ruff rule set is finalised as follows:
```
"E", # pycodestyle
"F", # Pyflakes
"UP", # pyupgrade
"I", # isort
"NPY", # Numpy
"PERF", # Perflint
"LOG", # flake8-logging
"ICN", # flake8-import-conventions
"SIM", # flake8-simplify
```

**Note:** There are more rule sets that will be useful to add down the line. They are not added in this PR to limit the scope but I will create some issues for those rule sets.

<!---

Describe your changes and why you're making them.

-->


## Testing Instructions

<!---

Please give any specific testing instructions to the reviewer here.

-->

1. run `pip install ruff` or `pip install --upgrade .[dev]` in an existing environment or `pip install .[dev]` in a new env.
2. run `ruff check .`. It should say **All checks passed!**. But incase of errors,  run `ruff check --fix` to fix the fixable mistakes. The rest will have to be manually fixed.
3. run `ruff format --check`. It should say **153 files already formatted**. But incase of formatting errors, run `ruff format .`

## Checklist

<!---

This checklist is mostly useful as a reminder of small things that can easily be

forgotten – it is meant as a helpful tool rather than hoops to jump through.

Put an `x` in all the items that apply, make notes next to any that haven't been

addressed, and remove any items that are not relevant to this PR.

-->

- [x] The title is of a format appropriate for a line in future release notes.
- [x] I have added the appropriate tags to the PR.
- [x] My PR represents one logical piece of work.
- [x] My PR fully fixes the issue linked. If new issues have been created, what are they?
      - #1326 
      - #1327
      - #1328
      - #1329
      - #1330
- [x] I have added the appropriate tests to cover code that has been added.
- [x] I have updated the documentation in the relevant places to cover the changes.

AndrewLister-STFC reviewed (commented):
I've had a look through some of this. It might take a few sessions to skim though all 142 files... all looks good so far.
I put a few questions in, and stuck some emojis on changes I thought were particularly good or bad.

Overall I think the trade-off is absolutely worth it.

AndrewLister-STFC reviewed (commented):
I've had a look through some of this. It might take a few sessions to skim though all 142 files... all looks good so far.
I put a few questions in, and stuck some emojis on changes I thought were particularly good or bad.

Overall I think the trade-off is absolutely worth it.

RabiyaF commented on fitbenchmarking/cli/tests/test_checkpoint_handler.py:
```diff
@@ -237,13 +262,10 @@ def setUp(self):
-        self.C = json.load((ch_test_files/'C.json').open()
-                           )['DataSet1']['problems']
+        ch_test_files = Path(__file__).parent / "test_files"
+        self.A = json.load((ch_test_files / "A.json").open())["DataSet1"]["problems"]
```
Originally, flake-8 was 79 and then I changed it to 90 and then checked the default length in ruff. Which is 88 (I believe). I did not have a strong reason to stick with non default value so, I just removed it.

AndrewLister-STFC replied:
No strong opinions from me. We were using the pep8 suggestion https://peps.python.org/pep-0008/#maximum-line-length
I'm not sure it matters too much as long as it's not wildly different from that

RabiyaF replied:
I will add this line back into the [ruff.lint] settings. The first commit should have this line length. If that looks good then this is an easy one line change.
<img width="266" alt="image" src="https://github.com/user-attachments/assets/8d49ec02-608f-4401-bf33-55f64a23de05">

RabiyaF replied:
Let's go with 79 because we were not using black and it would be consistent with the original flake8 configuration.

<img width="965" alt="Screenshot 2024-10-11 at 11 47 01" src="https://github.com/user-attachments/assets/24860c8e-c889-4545-9167-8cac7eea4e36">

RabiyaF commented on fitbenchmarking/cli/tests/test_checkpoint_handler.py:
```diff
@@ -84,98 +92,114 @@ def setUp(self):
 
-        assert not errors, 'Files do not match. ' \
+        assert not errors, (
+            'Files do not match. '
```
Yes I wanted to do this. It has automatically updated most of them but not all of them. I manually went through the files and tried to correct some. You will notice some discrepancies in the html tags as well.

RabiyaF commented on fitbenchmarking/cli/tests/test_checkpoint_handler.py:
```diff
@@ -84,98 +92,114 @@ def setUp(self):
         merge_data_sets([self.A, self.B], str(intermediate.absolute()))
-        twomerge = self.dir / 'two-merges.json'
+        twomerge = self.dir / "two-merges.json"
         merge_data_sets([str(intermediate), self.C], str(twomerge))
-        onemerge = self.dir / 'one-merge.json'
+        onemerge = self.dir / "one-merge.json"
         merge_data_sets([self.A, self.B, self.C], str(onemerge))
 
         errors = []
-        for i, (l1, l2) in enumerate(zip(onemerge.read_text().splitlines(),
-                                         twomerge.read_text().splitlines())):
+        for i, (l1, l2) in enumerate(
+            zip(
+                onemerge.read_text().splitlines(),
+                twomerge.read_text().splitlines(),
+            )
+        ):
```
See `test_options_minimizers.py` 😆 

<img width="684" alt="image" src="https://github.com/user-attachments/assets/47585ed2-45b1-4f4c-ba37-5620253bf73b">

RabiyaF commented:
I totally understand. I went through them one by one before requesting a review. Most of the changes are in the line length, quotation marks and simplifying logic. I want to add more rules but after this PR is reviewed and merged. It was getting unmanageable for me as well. Take your time :)

I don't think it is a priority but I just wanted to get this in review. I will open a PR for more mantid data this week. 

There is another thing that I haven't opened an issue for yet but we can discuss in the meeting tomorrow. The typing is so inconsistent throughout the repo. I wanted to remove the quotation marks around the typing stuff and started throwing errors and I had to revert changes.

AndrewLister-STFC commented:
> I totally understand. I went through them one by one before requesting a review. Most of the changes are in the line length, quotation marks and simplifying logic. I want to add more rules but after this PR is reviewed and merged. It was getting unmanageable for me as well. Take your time :)
> 
> I don't think it is a priority but I just wanted to get this in review. I will open a PR for more mantid data this week.

👍 

> There is another thing that I haven't opened an issue for yet but we can discuss in the meeting tomorrow. The typing is so inconsistent throughout the repo. I wanted to remove the quotation marks around the typing stuff and started throwing errors and I had to revert changes.

I know I used the following pattern in some places. These imports don't need to exist for the code to run, just when doing the typing so it made sense to me to do it that way. I think it's also possible to have circular imports too if everything is strongly typed, although I don't know if this pattern would fix that.
```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import other_module

def foo(bar: 'other_module.SomeClass') -> None:
    ...
```

To make it work without any quotes, it's just editing the import list in the offending files.

RabiyaF commented on README.md:
```diff
@@ -4,7 +4,7 @@
 [![Coverage Status](https://img.shields.io/coveralls/github/fitbenchmarking/fitbenchmarking.svg?style=flat-square)](https://coveralls.io/github/fitbenchmarking/fitbenchmarking)
 [![Chat](https://img.shields.io/badge/chat-CompareFitMinimizers-lightgrey.svg?style=flat-square&logo=slack)](https://slack.com/)
-[![Zenodo](https://img.shields.io/badge/Zenodo-10.5281/zenodo.6597790-blue.svg?style=flat-square)](https://doi.org/10.5281/zenodo.6597790)
+[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.11198140.svg)](https://doi.org/10.5281/zenodo.11198140)
```
I wanted to sneak this one in with the rest of the changes. I noticed this when I added the install badge. The current badge was redirecting to an older version and the page then says a newer version of the software is available. I also noticed that Letizia's name is not in the CITATION.cff.

RabiyaF commented on docs/source/contributors/guidelines.rst:
```diff
@@ -19,9 +19,8 @@ The workflow to be used for submitting new code/issues is described in
-All pull requests should be `PEP 8 compliant <https://www.python.org/dev/peps/pep-0008/>`_.
-We suggest running code through `flake8 <https://flake8.pycqa.org/en/latest/>`_ and
-`pylint <https://www.pylint.org/>`_ before submitting to check for this.
+All pull requests should be compliant with selected `Ruff<https://docs.astral.sh/ruff/>`_ rules.
```
Not yet.I have not added that ruleset. I created an issue for this #1329. It will make the variable naming compliant. But there will need to be about a 100 changes. Our repo is not compliant yet. So, I just removed that because we were not checking for it explicitly.

AndrewLister-STFC commented on fitbenchmarking/controllers/horace_controller.py:
```diff
@@ -64,49 +67,46 @@ def setup(self):
-                self.eng.evalc('horace_off;')
-            except matlab.engine.MatlabExecutionError:
-                pass
+        if self.cost_func.problem.format != "horace":
+            with contextlib.suppress(matlab.engine.MatlabExecutionError):
+                self.eng.evalc("horace_off;")
```
👍

AndrewLister-STFC reviewed (commented):
I've looked through and commented on 64 of the 142 files

jess-farmer commented on fitbenchmarking/cli/checkpoint_handler.py:
```diff
@@ -27,84 +31,105 @@ def get_parser() -> ArgumentParser:
-             '`fitbenchmarking-cp ACTION -h`')
+        metavar="ACTION",
+        dest="subprog",
+        help="Which action should be performed? "
+        "For more information on options use "
+        "`fitbenchmarking-cp ACTION -h`",
```
I agree, I think it is now less readable

AndrewLister-STFC commented on fitbenchmarking/results_processing/base_table.py:
```diff
@@ -438,16 +458,25 @@ def to_html(self):
         column_dividers = column_dividers[1:]
 
         # Set the cell colours and increase bars between cost functions
-        table_style = table.style\
-            .apply(lambda df: self.get_colour_df(like_df=df), axis=None)\
-            .set_table_styles(table_styles={
-                k: [{'selector': 'td',
-                     'props': [('border-left-width', '3px')]},
-                    {'selector': 'th',
-                     'props': [('border-left-width', '3px')]}]
-                for k in column_dividers})
+        table_style = table.style.apply(
+            lambda df: self.get_colour_df(like_df=df), axis=None
+        ).set_table_styles(
+            table_styles={
```
Again, here I prefer the old one 🤷

AndrewLister-STFC reviewed (commented):
120/164 files reviewed!

AndrewLister-STFC commented on fitbenchmarking/utils/checkpoint.py:
```diff
@@ -225,81 +235,82 @@ def load(self):
-        for f in [self.options.checkpoint_filename,
-                  os.path.join(self.options.results_dir,
-                               self.options.checkpoint_filename)]:
+        output: Dict[str, list[FittingResult]] = {}
```
Why is this `Dict[str, list[...]]` and not `Dict[str, List[...]]` or `dict[str, list[...]]`?

I think all 3 are correct since 3.9 (https://peps.python.org/pep-0585/). Is it an automated ruff change (i.e. we don't know why) or one that you put in to fix something?

AndrewLister-STFC replied:
Thinking about this a little more, taking out the quotes probably led to syntax errors in the 3.8 tests for the 3rd option since that pep only came in in 3.9

RabiyaF replied:
I think that in python 3.8, we can't use built in types for type checking. Once we update the python to 3.9 we can use `dict[str, list[...]]`. That would be the correct way to do it. But I want to create another issue to make typing consistent throughout the repo.

RabiyaF replied:
There is a GitHub action that failed because I updated it to `dict[str, list[...]]` I was trying to find it but that is why I have revert the typing changes.

AndrewLister-STFC commented on fitbenchmarking/utils/fitbm_result.py:
```diff
@@ -1,13 +1,17 @@
 import numpy as np
 from scipy import stats
 
+from fitbenchmarking.controllers.base_controller import Controller
+from fitbenchmarking.cost_func.base_cost_func import CostFunc
 from fitbenchmarking.cost_func.nlls_base_cost_func import BaseNLLSCostFunc
+from fitbenchmarking.parsing.fitting_problem import FittingProblem
```
If these are moving here, we can remove the `TYPE_CHECKING` guarded imports below.

AndrewLister-STFC reviewed (commented):
Ok, that's all my comments! Let me know if you want to chat about any of the comments I made.

RabiyaF commented:
Perfect. I will look at this next week and try to address the comments and will let you know if anything is unclear. I want to review Jess's PR first. So, the earliest I can look at is next week.

jess-farmer commented:
I've now had a look through all 142 files - I agree with Andrew that there are a few places where the changes have made things more difficult to read, however I don't think it's that bad and think that the positives of switching to Ruff make it worth it. I also get the expected output when running `ruff check` and `ruff format --check`, and the pre-commit hook works for me

Many thanks @RabiyaF for taking the time to make these changes!

RabiyaF commented:
I have addressed all the comments. One change I made was I removed the ruff version pin. This is because when I started this PR `v0.6.4` was the latest version and now `v.0.6.9` is the latest version. It is being developed rapidly and we should take advantage of the additional rules and fixes being added. However, the version in the precommit hook does need to be specified. So, I left that as `v0.6.9`.

AndrewLister-STFC commented on fitbenchmarking/parsing/tests/test_parsers.py:
```diff
@@ -382,11 +397,11 @@ def test_sparsej_evaluation(self, file_format, evaluations_file):
-            message = 'No function evaluations provided to test ' \
-                f'against for {file_format}'
-            assert (evaluations_file is not None), message
+            message = f"""No function evaluations provided 
+            to test against for {file_format}"""
```
```suggestion
            message = (
                "No function evaluations provided "
                f"to test against for {file_format}"
            )
```

AndrewLister-STFC commented on fitbenchmarking/utils/matlab_engine.py:
```diff
@@ -14,9 +15,8 @@ def add_persistent_matlab_var(name: str) -> None:
-        "if ~exist('persistent_vars');"
-        " persistent_vars = {'persistent_vars'};"
-        "end;"
+        """if ~exist('persistent_vars'); 
+        persistent_vars = {'persistent_vars'};end;"""
```
```suggestion
        "if ~exist('persistent_vars'); "
        "persistent_vars = {'persistent_vars'};end;"
```

AndrewLister-STFC commented on fitbenchmarking/utils/tests/test_misc.py:
```diff
@@ -23,19 +23,20 @@ def setUp(self):
-                        'is correct')
+            with open(filepath, "w+", encoding="utf-8") as f:
+                f.write(
+                    """This is a mock data file to check 
+                    that finding files is correct"""
```
```suggestion
                    "This is a mock data file to check "
                    "that finding files is correct"
```

AndrewLister-STFC commented on fitbenchmarking/results_processing/tables.py:
```diff
@@ -118,38 +133,42 @@ def create_results_tables(options, results, best_results, group_dir, fig_dir,
-                                    f'pp/{"+".join(table.pps)}',
+                            for p in table.pps
+                        ],
+                        pp_dash_url=f"""
+                        http://127.0.0.1:{options.port}/
+                        {os.path.basename(group_dir)}/pp/
+                        {'+'.join(table.pps)}""",
```
```suggestion
                        pp_dash_url=(
                            f"http://127.0.0.1:{options.port}/"
                            f"{os.path.basename(group_dir)}/pp/"
                            f"{'+'.join(table.pps)}"
                        ),
```

AndrewLister-STFC reviewed (changes requested):
I think I've looked through all of the new changes. I've put suggestions in for all of the triple quotes I could see to avoid the whitespace that triple quotes preserve. Let me know if you want to chat about any of them

RabiyaF commented:
I think all the changes from review 2 have now been applied. The failing tests are because of `levmar`, `gofit` not being installable on `3.12`. And sasview/sasmodels is using `np.NaN` in 7 files that is not compatible with numpy 2.0.

<img width="1228" alt="image" src="https://github.com/user-attachments/assets/97d09de2-9312-4c68-ada6-b063f5a3d69f">

AndrewLister-STFC commented on fitbenchmarking/results_processing/tables.py:
```diff
@@ -118,39 +133,42 @@ def create_results_tables(options, results, best_results, group_dir, fig_dir,
                         pp_dash_url=f'http://127.0.0.1:{options.port}/'
-                                    f"{options.run_id}/"
-                                    f'{os.path.basename(group_dir)}/'
-                                    f'pp/{"+".join(table.pps)}',
+                        f"{options.run_id}/"
+                        f'{os.path.basename(group_dir)}/'
+                        f'pp/{"+".join(table.pps)}',
```
I'd suggest this here:

```
pp_dash_url=(
    f"http://127.0.0.1:{options.port}/"
    f"{options.run_id}/"
    f"{os.path.basename(group_dir)}/"
    f"pp/{'+'.join(table.pps)}"
),

AndrewLister-STFC reviewed (commented):
I only have 1 last comment. Seems like 1 got through the net when rebasing

AndrewLister-STFC reviewed (commented):
I only have 1 last comment. Seems like 1 got through the net when rebasing

AndrewLister-STFC reviewed (approved):
Looks good to me!!

jess-farmer reviewed (approved):
Looks great - thanks Rabiya!
